File: Tree/ BinaryTree/ BinaryTree.cpp
#include <iostream>

class TreeNode {
public:
    int key;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int k) : key(k), left(nullptr), right(nullptr) {}
};

class BinaryTree {
private:
    TreeNode* root;

    TreeNode* insert(TreeNode* node, int key) {
        if (node == nullptr)
            return new TreeNode(key);

        if (key < node->key)
            node->left = insert(node->left, key);
        else if (key > node->key)
            node->right = insert(node->right, key);

        return node;
    }

    TreeNode* remove(TreeNode* node, int key) {
        if (node == nullptr)
            return node;

        if (key < node->key)
            node->left = remove(node->left, key);
        else if (key > node->key)
            node->right = remove(node->right, key);
        else {
            if (node->left == nullptr) {
                TreeNode* temp = node->right;
                delete node;
                return temp;
            } else if (node->right == nullptr) {
                TreeNode* temp = node->left;
                delete node;
                return temp;
            }

            node->key = getMinValue(node->right);
            node->right = remove(node->right, node->key);
        }
        return node;
    }

    int getMinValue(TreeNode* node) {
        while (node->left != nullptr)
            node = node->left;
        return node->key;
    }

    TreeNode* search(TreeNode* node, int key) {
        if (node == nullptr || node->key == key)
            return node;
        if (key < node->key)
            return search(node->left, key);
        return search(node->right, key);
    }

    void inOrderTraversal(TreeNode* node) {
        if (node != nullptr) {
            inOrderTraversal(node->left);
            std::cout << node->key << " ";
            inOrderTraversal(node->right);
        }
    }

public:
    BinaryTree() : root(nullptr) {}

    void insert(int key) {
        root = insert(root, key);
    }

    void remove(int key) {
        root = remove(root, key);
    }

    bool search(int key) {
        return search(root, key) != nullptr;
    }

    void printInOrderTraversal() {
        inOrderTraversal(root);
    }
};

int main() {
    BinaryTree bst;
    bst.insert(10);
    bst.insert(5);
    bst.insert(15);
    bst.insert(3);
    bst.insert(7);

    std::cout << "Inorder traversal: ";
    bst.printInOrderTraversal();
    std::cout << std::endl;

    bst.remove(5);
    std::cout << "After deleting 5: ";
    bst.printInOrderTraversal();
    std::cout << std::endl;

    bool searchResult = bst.search(10);
    if (searchResult)
        std::cout << "Node with key 10 found." << std::endl;
    else
        std::cout << "Node with key 10 not found." << std::endl;

    return 0;
}

File: Tree/ BinaryTree/ BinaryTree.py
class TreeNode:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self):
        self.root = None

    def insert(self, key):
        self.root = self._insert(self.root, key)

    def _insert(self, root, key):
        if root is None:
            return TreeNode(key)

        if key < root.key:
            root.left = self._insert(root.left, key)
        elif key > root.key:
            root.right = self._insert(root.right, key)

        return root

    def delete(self, key):
        self.root = self._delete(self.root, key)

    def _delete(self, root, key):
        if root is None:
            return root

        if key < root.key:
            root.left = self._delete(root.left, key)
        elif key > root.key:
            root.right = self._delete(root.right, key)
        else:
            if root.left is None:
                return root.right
            elif root.right is None:
                return root.left

            root.key = self._get_min_value(root.right)
            root.right = self._delete(root.right, root.key)

        return root

    def _get_min_value(self, root):
        while root.left is not None:
            root = root.left
        return root.key

    def search(self, key):
        return self._search(self.root, key)

    def _search(self, root, key):
        if root is None or root.key == key:
            return root
        if key < root.key:
            return self._search(root.left, key)
        return self._search(root.right, key)

    def inorder_traversal(self):
        result = []
        self._inorder_traversal(self.root, result)
        return result

    def _inorder_traversal(self, root, result):
        if root is not None:
            self._inorder_traversal(root.left, result)
            result.append(root.key)
            self._inorder_traversal(root.right, result)

bst = BinaryTree()
bst.insert(10)
bst.insert(5)
bst.insert(15)
bst.insert(3)
bst.insert(7)

print("Inorder traversal:", bst.inorder_traversal())

bst.delete(5)
print("After deleting 5:", bst.inorder_traversal())

search_result = bst.search(10)
if search_result:
    print("Node with key 10 found.")
else:
    print("Node with key 10 not found.")